src/_pytest/_raises_group.py

```diff
@@ -0,0 +1,1003 @@
+    if expected_type is None:
+        return None
+
+    if not isinstance(
+        exception,
+        expected_type,
+    ):
+        actual_type_str = _exception_type_name(type(exception))
+        expected_type_str = _exception_type_name(expected_type)
+        if isinstance(exception, BaseExceptionGroup) and not issubclass(
+            expected_type, BaseExceptionGroup
+        ):
+            return f"Unexpected nested {actual_type_str}, expected {expected_type_str}"
+        return f"{actual_type_str} is not of type {expected_type_str}"
+    return None
+
+
+class AbstractMatcher(ABC, Generic[BaseExcT_co]):
+    """ABC with common functionality shared between Matcher and RaisesGroup"""
+
+    def __init__(
+        self,
+        match: str | Pattern[str] | None,
+        check: Callable[[BaseExcT_co], bool] | None,
+    ) -> None:
+        if isinstance(match, str):
+            self.match: Pattern[str] | None = re.compile(match)
+        else:
+            self.match = match
+        self.check = check
+        self._fail_reason: str | None = None
+
+        # used to suppress repeated printing of `repr(self.check)`
+        self._nested: bool = False
+
+    @property
+    def fail_reason(self) -> str | None:
+        """Set after a call to `matches` to give a human-readable reason for why the match failed.
+        When used as a context manager the string will be given as the text of an
+        `Failed`"""
+        return self._fail_reason
+
+    def _check_check(
+        self: AbstractMatcher[BaseExcT_1],
+        exception: BaseExcT_1,
+    ) -> bool:
+        if self.check is None:
+            return True
+
+        if self.check(exception):
+            return True
+
+        check_repr = "" if self._nested else " " + repr_callable(self.check)
+        self._fail_reason = f"check{check_repr} did not return True"
+        return False
+
+    def _check_match(self, e: BaseException) -> bool:
+        if self.match is None or re.search(
+            self.match,
+            stringified_exception := _stringify_exception(e),
+        ):
+            return True
+
+        maybe_specify_type = (
+            f" of {_exception_type_name(type(e))}"
+            if isinstance(e, BaseExceptionGroup)
+            else ""
+        )
+        self._fail_reason = (
+            f"Regex pattern {_match_pattern(self.match)!r}"
+            f" did not match {stringified_exception!r}{maybe_specify_type}"
+        )
+        if _match_pattern(self.match) == stringified_exception:
+            self._fail_reason += "\n  Did you mean to `re.escape()` the regex?"
+        return False
+
+    @abstractmethod
+    def matches(
+        self: AbstractMatcher[BaseExcT_1], exc_val: BaseException
+    ) -> TypeGuard[BaseExcT_1]:
+        """Check if an exception matches the requirements of this AbstractMatcher.
+        If it fails, `AbstractMatcher.fail_reason` should be set.
+        """
+
+
+@final
+class Matcher(AbstractMatcher[MatchE]):
```

Comment:
We ought not to use such a generic name easily 

There's a general wishlist plan to enable matchers as a whole that act similar to dirty equals 

The matcher api that's being used here conflates multiple aspects of that in a manner that's problematic to partition later


Ideally the whole api around matchers is first evolved as a plugin to avoid early api lock in as it's hard to make good matchers 

I'll elaborate further when the flu is over

Reply:
sure, I can rename it... maybe something that's closer to `RaisesGroup`... `RaisesExc`/`RaisesException`? or something like `ExpectedExc`/`ExpectedException`

This has been available & in use for >1 year in trio, which has allowed the interface to mature. I don't think pytest needs to repeat that process by putting it in a plugin.

I know you've mentioned matcher objects in #11538 and I'm excited for what that might entail in the future, but I don't think we want to wait for that to add usable exceptiongroup support. If you have concrete ways of improving the `RaisesGroup` interface to harmonize with it I'm of course open to suggestions though.